### PR TITLE
feat: add transcript c apis for kroma prover

### PR DESCRIPTION
**Pull request by chokobole**
# Description

To run [kroma-prover v0.1.4](https://github.com/kroma-network/kroma-prover/tree/v0.1.4), we need to support more transcripts, `Poseidon` and `Sha256`. This PR creates c APIs for them.

### CODE REVIEW

File: tachyon/c/zk/plonk/halo2/bn254_transcript.h [hunk 1]
```diff
@@ -33,10 +34,18 @@ TACHYON_C_EXPORT void tachyon_halo2_bn254_transcript_writer_update(
...
 // If |data| is NULL, then it populates |data_len| with length to be used.
 // If |data| is not NULL, then it populates |data| with the hash.
+// If the type of the transcript is poseidon, you should call
+// |tachyon_halo2_bn254_transcript_writer_squeeze| instead.
+// Otherwise, it terminates the program.
 TACHYON_C_EXPORT void tachyon_halo2_bn254_transcript_writer_finalize(
     tachyon_halo2_bn254_transcript_writer* writer, uint8_t* data,
     size_t* data_len);
 
+// If the type of the transcript is other than poseidon, it terminates the
```
Comment (ashjeong) [inline tachyon/c/zk/plonk/halo2/bn254_transcript.h:44]:
```suggestion
// If the type of the transcript is not poseidon, it terminates the
```

**Review by ashjeong — COMMENTED** [review]
In previous commits like 5e12d2143c238789d6f3d8807b07e816329483e7 , ec46ffc000cf0432044556b5d97476f0a849c8a5 , and ced3d58bc0bf30609e757e489e50c3400d594b57 you include the copyable unit test in the commit about adding copyable, but you don’t do so for f73a8d5acac3ecd016c49d6e1a803508dddb6196 . Can you change this commit to be part of https://github.com/kroma-network/tachyon/pull/300/commits/ced3d58bc0bf30609e757e489e50c3400d594b57 or separate the other commits as well?

**Comment by chokobole** [comment]
> In previous commits like [5e12d21](https://github.com/kroma-network/tachyon/commit/5e12d2143c238789d6f3d8807b07e816329483e7) , [ec46ffc](https://github.com/kroma-network/tachyon/commit/ec46ffc000cf0432044556b5d97476f0a849c8a5) , and [ced3d58](https://github.com/kroma-network/tachyon/commit/ced3d58bc0bf30609e757e489e50c3400d594b57) you include the copyable unit test in the commit about adding copyable, but you don’t do so for [f73a8d5](https://github.com/kroma-network/tachyon/commit/f73a8d5acac3ecd016c49d6e1a803508dddb6196) . Can you change this commit to be part of [ced3d58](https://github.com/kroma-network/tachyon/commit/ced3d58bc0bf30609e757e489e50c3400d594b57) or separate the other commits as well?


I intentionally structured it this way because to add a Copyable unittest, I needed to implement a commit that adds the equality operator first.

1. Equality operator
2. Copyable
3. Copyable test

However, upon reflection, I realize I can combine the implementation and test commits into one by changing the order as follows:

1. Equality operator
2. Copyable

**Review by ashjeong — APPROVED** [review]
LGTM

File: tachyon/c/zk/plonk/halo2/bn254_transcript.cc [hunk 2]
```diff
@@ -85,8 +109,28 @@ void tachyon_halo2_bn254_transcript_writer_finalize(
...
+
+tachyon_bn254_fr tachyon_halo2_bn254_transcript_writer_squeeze(
+    tachyon_halo2_bn254_transcript_writer* writer) {
+  tachyon_bn254_fr ret;
+  math::bn254::Fr challenge;
+  switch (static_cast<zk::plonk::halo2::TranscriptType>(writer->type)) {
+    case zk::plonk::halo2::TranscriptType::kPoseidon: {
+      challenge = reinterpret_cast<PoseidonWriter*>(writer->extra)->Squeeze();
+      memcpy(ret.limbs, challenge.value().limbs,
+             math::bn254::Fr::BigIntTy::kByteNums);
```
Comment (TomTaehoonKim) [inline tachyon/c/zk/plonk/halo2/bn254_transcript.cc:155]:
```suggestion
             math::bn254::Fr::BigIntTy::kLimbNums);
```

**Review by TomTaehoonKim — CHANGES_REQUESTED** [review]

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by Insun35 — APPROVED** [review]
LGTM

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Pull request merged**